File: lib/models/get_home_location.dart
import 'package:geolocator/geolocator.dart';

/// GetHomeLocation is a class that represents a location with latitude and longitude.
/// It's goal is to provide a generic way to handle locations in the app and standardize datatypes
class GetHomeLocation {
  final double _lat;
  final double _lng;

  /// Constructor:Creates a GetHomeLocation with the given latitude and longitude.
  GetHomeLocation({required double lat, required double lng}) : _lat = lat, _lng = lng;

  /// Creates a GetHomeLocation from a Position object (from package:geolocator_platform_interface)
  factory GetHomeLocation.fromPosition(Position position) {
    return GetHomeLocation(lat: position.latitude, lng: position.longitude);
  }

  /// Creates an empty GetHomeLocation with lat and lng set to 0.
  factory GetHomeLocation.empty() {
    return GetHomeLocation(lat: 0, lng: 0);
  }


  /// Converts a JSON object to a GetHomeLocation
  factory GetHomeLocation.fromJson(Map<String, dynamic> json) {
    return GetHomeLocation(
      lat: json['lat'],
      lng: json['lng'],
    );
  }

  /// Converts a GetHomeLocation to a JSON-encodable map (default call by dart:convert.jsonEncode()).
  Map<String, dynamic> toJson() {
    return {
      'lat': _lat,
      'lng': _lng,
    };
  }

  // Returns the distance in km
  double getDistanceTo(GetHomeLocation location) {
    return Geolocator.distanceBetween(_lat, _lng, location._lat, location._lng);
  }

  // Returns true if the location is empty (lat and lng are 0)
  bool isEmpty() {
    return _lat == 0 && _lng == 0;
  }

  // Getters
  double getLatitude() => _lat;
  double getLongitude() => _lng;

  /// toString method
  @override
  String toString() {
    return 
      '\n GetHomeLocation:'
      '\n Latitude: $_lat,'
      '\n Longitude: $_lng'
      '\n';
  }
}

File: ios/GetHome Widget/get-home-location.swift
//
//  get-home-location.swift
//  Runner
//
//  Created by Lennart Hesse on 20.04.24.
//

import Foundation

struct GetHomeLocation {
    let lat: Double
    let lng: Double

    init(lat: Double, lng: Double) {
        self.lat = lat
        self.lng = lng
    }

    static func fromJson(json: [String: Any]) -> GetHomeLocation? {
        guard let lat = json["lat"] as? Double, let lng = json["lng"] as? Double else {
            return nil
        }
        return GetHomeLocation(lat: lat, lng: lng)
    }

    func toJson() -> [String: Any] {
        return ["lat": lat, "lng": lng]
    }

    var latitude: Double {
        return lat
    }

    var longitude: Double {
        return lng
    }

    var description: String {
        return """
        GetHomeLocation:
        Latitude: \(lat)
        Longitude: \(lng)
        """
    }
}
